src/Relude/List.hs

```diff
@@ -13,7 +13,34 @@ Utility functions to work with lists.
 module Relude.List
        ( module Relude.List.NonEmpty
        , module Relude.List.Reexport
+       , (!!?)
        ) where
 
+import Data.Bool (otherwise, (||))
+import Data.List (length, (!!))
+import Data.Maybe (Maybe (..))
+
+import Relude.Base ((<), (>=))
 import Relude.List.NonEmpty
 import Relude.List.Reexport
+import Relude.Numeric (Int)
+
+-- $setup
+-- >>> import Relude
+{- | Safer version of '!!', returns a Maybe.
+get element from list using index value starting from `0`.
+
+>>> [] !!? 0
+Nothing
+
+>>> ["a", "b", "c"] !!? 3
+Nothing
+
+>>> ["a", "b", "c"] !!? 2
+Just "c"
+-}
+(!!?) :: [a] -> Int -> Maybe a
+(!!?) xs i
+  | i < 0 || i >= length xs = Nothing
```

Comment:
Checking the `length` is not very performant. I think it's better to write a recursive implementation by hands instead of reusing `!!`.